docs/design/phases.md

```diff
@@ -0,0 +1,354 @@
+ - Any extra flags and/or HostPath Volumes/VolumeMounts specified by the user
+
+Other flags that are set unconditionally:
+ - The `BootstrapTokenAuthenticator` authentication module is enabled
+ - `--client-ca-file` to `ca.crt`
+ - `--tls-cert-file` to `apiserver.crt`
+ - `--tls-private-key-file` to `apiserver.key`
+ - `--kubelet-client-certificate` to `apiserver-kubelet-client.crt`
+ - `--kubelet-client-key` to `apiserver-kubelet-client.key`
+ - `--service-account-key-file` to `sa.pub`
+ - `--requestheader-client-ca-file` to `front-proxy-ca.crt`
+ - `--admission-control` to `NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,DefaultStorageClass,ResourceQuota` or whatever the recommended set of admission controllers is at the time
+ - `--storage-backend` to `etcd3`. Support for `etcd2` in kubeadm is dropped.
+ - `--kubelet-preferred-address-types` to `InternalIP,ExternalIP,Hostname` so `kubectl logs` and other apiserver -> kubelet communication works in environments where the hostnames of the nodes aren't resolvable.
+ - `--requestheader-username-headers=X-Remote-User`, `--requestheader-group-headers=X-Remote-Group`, `--requestheader-extra-headers-prefix=X-Remote-Extra-`, --requestheader-allowed-names=front-proxy-client` so the front proxy communication is secure.
+
+
+#### controller-manager
+
+The controller-manager needs to know this in particular:
+ - The Pod Network CIDR if any; also enables the Subnet Manager feature (required for some CNI network plugins)
+ - If auto-approval of CSRs should be enabled
+
+Other flags that are set unconditionally:
+ - The `BootstrapSigner` and `TokenCleaner` controllers are enabled
+ - `--root-ca-file` to `ca.crt`
+ - `--cluster-signing-cert-file` to `ca.crt`
+ - `--cluster-signing-key-file` to `ca.key`
+ - `--service-account-private-key-file` to `sa.key`
+ - `--use-service-account-credentials` to true
+
+#### Scheduler
+
+kubeadm doesn't set any special flags to the scheduler
+
+Common properties for the control plane components:
+ - Leader election is enabled for both the controller-manager and the scheduler
+ - `HostNetwork: true` is present on all static pods since there is no network configured yet
+ - The yaml manifests are located in `/etc/kubernetes/manifests`
+
+### Addons
+
+#### mark-master
+
+This addon essentially does just this:
+
+```bash
+kubectl taint node ${master_name} node-role.kubernetes.io/master:NoSchedule
+kubectl label node ${master_name} node-role.kubernetes.io/master=""
+```
+
+#### cluster-info
+
+This phase creates the `cluster-info` ConfigMap in the `kube-public` namespace as defined in https://github.com/kubernetes/community/blob/master/contributors/design-proposals/bootstrap-discovery.md
+ - The `ca.crt` and the address/port of the apiserver is added to the `cluster-info` ConfigMap in the `kubeconfig` key
+ - Exposes the `cluster-info` ConfigMap to unauthenticated users (i.e. users in RBAC group `system:unauthenticated`)
+
```

Comment:
Looking at `kubeadm init` code, it seems there are also two additional steps here: `CreateServiceAccounts` and `CreateRBACRules`.
In order to make `kubeadm init` a full "bundle command", also those steps should be implemented as phases.